### PR TITLE
"add dynamic lstm scripts"

### PR DESCRIPTION
**Pull request by dzhwinter**
fix https://github.com/PaddlePaddle/Paddle/issues/6165

### CODE REVIEW

File: paddle/understand_sentiment_dynamic_lstm.py [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    fc1 = fluid.layers.fc(input=emb, size=hid_dim)
+    lstm1, cell1 = fluid.layers.dynamic_lstm(input=fc1, size=hid_dim)
+
+    inputs = [fc1, lstm1]
+
+    for i in range(2, stacked_num + 1):
+        fc = fluid.layers.fc(input=inputs, size=hid_dim)
+        lstm, cell = fluid.layers.dynamic_lstm(
+            input=fc, size=hid_dim, is_reverse=(i % 2) == 0)
+        inputs = [fc, lstm]
```
Comment (qingqing01) [inline paddle/understand_sentiment_dynamic_lstm.py]: The configuration here is not consistent with the config in https://github.com/dzhwinter/benchmark/pull/2/files . There is no reversed LSTM in that PR. Just simple stacked LSTM.

File: paddle/understand_sentiment_dynamic_lstm.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    start_time = time.time()
+    word_dict = paddle.dataset.imdb.word_dict()
+
+    print("load word dict successfully")
+
+    dict_dim = len(word_dict)
+    # data = fluid.layers.data(
+    #     name="words", shape=[1], append_batch_size=False, dtype="int64")
+    # label = fluid.layers.data(
+    #     name="label", shape=[1], append_batch_size=False, dtype="int64")
```
Comment (qingqing01) [inline paddle/understand_sentiment_dynamic_lstm.py]: Remove the unused codes.

File: paddle/understand_sentiment_dynamic_lstm.py [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        default='CPU',
+        choices=['CPU', 'GPU'],
+        help='The device type.')
+    parser.add_argument(
+        '--infer_only', action='store_true', help='If set, run forward only.')
+    parser.add_argument(
+        '--use_cprof', action='store_true', help='If set, use cProfile.')
+    parser.add_argument(
+        '--use_nvprof',
+        action='store_false',
```
Comment (qingqing01) [inline paddle/understand_sentiment_dynamic_lstm.py]: store_false -> store_true

Comment (dzhwinter) [reply paddle/understand_sentiment_dynamic_lstm.py]: fixed.

File: paddle/understand_sentiment_dynamic_lstm.py (hunk 1 above)
Comment (dzhwinter) [reply paddle/understand_sentiment_dynamic_lstm.py]: fixed

File: paddle/understand_sentiment_dynamic_lstm.py (hunk 2 above)
Comment (dzhwinter) [reply paddle/understand_sentiment_dynamic_lstm.py]: fixed.

File: paddle/understand_sentiment_dynamic_lstm.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    print('------------------------------------------------')
+
+
+def dynamic_lstm_model(data, dict_dim, class_dim=2):
+    batch_size = args.batch_size
+    emb_dim = args.emb_dim
+    hid_dim = args.hid_dim
+    stacked_num = args.stacked_num
+
+    assert stacked_num % 2 == 1, "Must stacked_num %2 == 1."
```
Comment (qingqing01) [inline paddle/understand_sentiment_dynamic_lstm.py]: remove this check.

File: paddle/understand_sentiment_dynamic_lstm.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+    fc1 = fluid.layers.fc(input=emb, size=hid_dim)
+    lstm1, cell1 = fluid.layers.dynamic_lstm(input=fc1, size=hid_dim)
+
+    inputs = [fc1, lstm1]
+
+    for i in range(stacked_num):
+        fc = fluid.layers.fc(input=inputs, size=hid_dim)
+        lstm, cell = fluid.layers.dynamic_lstm(input=fc, size=hid_dim)
+        inputs = [fc, lstm]
```
Comment (qingqing01) [inline paddle/understand_sentiment_dynamic_lstm.py]: inputs = lstm

line 66 67实际可以和line 71 - lin 74行合并。

File: paddle/understand_sentiment_dynamic_lstm.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    stacked_num = args.stacked_num
+
+    assert stacked_num % 2 == 1, "Must stacked_num %2 == 1."
+
+    emb = fluid.layers.embedding(input=data, size=[dict_dim, emb_dim])
+
+    fc1 = fluid.layers.fc(input=emb, size=hid_dim)
+    lstm1, cell1 = fluid.layers.dynamic_lstm(input=fc1, size=hid_dim)
+
+    inputs = [fc1, lstm1]
```
Comment (qingqing01) [inline paddle/understand_sentiment_dynamic_lstm.py]: inputs = lstm1

File: paddle/understand_sentiment_dynamic_lstm.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+    for i in range(stacked_num):
+        fc = fluid.layers.fc(input=inputs, size=hid_dim)
+        lstm, cell = fluid.layers.dynamic_lstm(input=fc, size=hid_dim)
+        inputs = [fc, lstm]
+
+    fc_last = fluid.layers.sequence_pool(input=inputs[0], pool_type='max')
+    lstm_last = fluid.layers.sequence_pool(input=inputs[1], pool_type='max')
```
Comment (qingqing01) [inline paddle/understand_sentiment_dynamic_lstm.py]: 只需要对最后的lstm进行pooling即可，line 79只有一个pooling 之后的输入。

File: paddle/understand_sentiment_dynamic_lstm.py [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    cost = fluid.layers.cross_entropy(input=prediction, label=label)
+    avg_cost = fluid.layers.mean(x=cost)
+    adam_optimizer = fluid.optimizer.Adam(learning_rate=0.002)
+    adam_optimizer.minimize(avg_cost)
+    accuracy = fluid.evaluator.Accuracy(input=prediction, label=label)
+
+    train_reader = paddle.batch(
+        paddle.reader.shuffle(
+            paddle.dataset.imdb.train(word_dict),
+            buf_size=args.batch_size * 10),
```
Comment (qingqing01) [inline paddle/understand_sentiment_dynamic_lstm.py]: 同https://github.com/dzhwinter/benchmark/pull/4#discussion_r154847252

**Comment by dzhwinter** [comment]
all fixed based on comments.

**Pull request merged**